### PR TITLE
doc: Batch proving in Circuits and L1

### PR DESCRIPTION
**Pull request by spalladino**
Related to AztecProtocol/aztec3-packages#7346

### CODE REVIEW

File: in-progress/7346-batch-proving-circuits-and-l1.md
Comment (alexghr) [inline in-progress/7346-batch-proving-circuits-and-l1.md:215]: Maybe we need another set of variables to track in-flight proving, otherwise publishing the 33rd block would overwrite the `last*` variables in the rollup which would fail proof validation.

Probably something to think about for proving coordination and having just a verified/last set of variables is enough.

**Review by alexghr — APPROVED** [review]
:fire:

**Review by iAmMichaelConnor — COMMENTED** [review]
I haven't finished, but here's some early comments. :)
Also, I drew this picture. Would this match your understanding? https://miro.com/app/board/uXjVK4BC8Yg=/ Notice also the blobification circuit, which I think should be loosely considered as part of this proposal. If you don't want to think about actual blobs, then perhaps we could think in terms of sha256-hashes, and at least contemplate what data needs to be hashed and emitted when we reach the 'block root' rollup circuit?
Is there any data above the "block" level (i.e. at the supra-block level) that needs to be submitted to blobs? If so, I'm missing yet another blobification circuit to feed the new 'root rollup' circuit.

(I'll be doing some thinking around this too)

File: in-progress/7346-batch-proving-circuits-and-l1.md
Comment (spalladino) [reply in-progress/7346-batch-proving-circuits-and-l1.md:146]: Yep, that's correct!

File: in-progress/7346-batch-proving-circuits-and-l1.md
Comment (spalladino) [reply in-progress/7346-batch-proving-circuits-and-l1.md:215]: > otherwise publishing the 33rd block would overwrite the last* variables in the rollup which would fail proof validation

Agree! That's why I suggested moving the "prove that this block root is in the current archive tree" to Solidity instead of the root rollup circuit. The proof would have a "last block proven", and the submitter would include a merkle path from it to the current `lastArchiveTreeRoot`. If that root changes, then the submitter needs to generate a new merkle path instead of rebuilding the entire root proof.

To further mitigate against this race condition, we could keep the last N archive tree roots instead of just the latest.

File: in-progress/7346-batch-proving-circuits-and-l1.md
Comment (spalladino) [reply in-progress/7346-batch-proving-circuits-and-l1.md:187]: That hackmd assumed we needed to create one proof per block, which is no longer the case, so we can replace "empty txs" with "empty blocks" in the superblock (name tbd!) proof.

File: in-progress/7346-batch-proving-circuits-and-l1.md
Comment (spalladino) [reply in-progress/7346-batch-proving-circuits-and-l1.md:230]: > The process name is unclear to me. Would this be better described as submitBlockProposal?

Depends on what we define as a "block proposal". Naming strikes again! If you mean "a block that is part of the unproven chain", then yeah.

> I think perhaps 2 (which will be calling the blob point eval precompile on some earlier-submitted blob) should be deferred until the proof is being verified.

Blobification is not yet clear to me, but yeah, we can defer that until the proof is submitted and verified if it makes sense. FWIW there's a step missing here that is checking the attestations from the validator committee, but that'll come in later.

> The L1-to-L2 messages shouldn't be removed from the L1 box until the proof has been verified

Good point! Do you know if we'd need to soft-delete them at least, in order to prevent them from being consumed in two unproven blocks in a row? Or is there another mechanism to prevent this?

> Validate the new block header - I thought we didn't want to submit potentially malicious state roots at the time of proposing a block? (and hence shouldn't be submitting new block headers to L1 at this stage)

This proposal is assuming we're still uploading tx effects, but yeah, this may change if/when we switch to tx objects.

**Comment by spalladino** [comment]
@iAmMichaelConnor that picture makes sense, yep! At first glance, blobification makes sense, and yeah, it should happen at the "block root" level.

> Is there any data above the "block" level (i.e. at the supra-block level) that needs to be submitted to blobs? If so, I'm missing yet another blobification circuit to feed the new 'root rollup' circuit.

Not that I can think of. The only things that happen at the root-root level are fee payment and L2-to-L1 messages, neither of which need blobspace I believe.

File: in-progress/7346-batch-proving-circuits-and-l1.md
Comment (LeilaWang) [reply in-progress/7346-batch-proving-circuits-and-l1.md]: > The L1-to-L2 messages shouldn't be removed from the L1 box until the proof has been verified

Was wondering about the same thing! I think we need to restore `toConsume` (but not `isProgress`) in the inbox to the point `verifiedArchiveTreeRoot` was set when a proof fails.

Was also considering deferring `l1_to_l2_roots` to be proved in the final root rollup and consume the root in the inbox from `submiteProof`. But this will result in a bigger delay for users to make use of a l1_to_l2_msg. So maybe not a good idea.

> This proposal is assuming we're still uploading tx effects, but yeah, this may change if/when we switch to tx objects.

What's tx objects?

**Review by PhilWindle — COMMENTED** [review]
Looks sensible to me. Definitely needs input from the other teams however.

File: in-progress/7346-batch-proving-circuits-and-l1.md
Comment (LHerskind) [reply in-progress/7346-batch-proving-circuits-and-l1.md:215]: A fairly simple approach that I think I would take for the first version at least, would be to simple store a "list" of the archive values instead of just the tips. 

This could be as:
```solidity
mapping(uint256 blockNumber => bytes32 archive) public archives;
```

Then you keep track of 2 indexes for how far we have added blocks, and how far we have proven.

This would also allow providing a proof that is not the "next", e.g., to avoid the issue where someone is delaying their proof to limit the time you have to interact. We are looking to add some of this in https://github.com/AztecProtocol/aztec-packages/issues/7614.

File: in-progress/7346-batch-proving-circuits-and-l1.md
Comment (LHerskind) [inline in-progress/7346-batch-proving-circuits-and-l1.md:232]: L1 to L2 messages don't need to be deleted from the inbox, they are not really deleted there, just progressing what tree we are reading. It still needs to perform some action at the "submitBlock", but we can get this fairly easily https://github.com/AztecProtocol/aztec-packages/issues/7592

File: in-progress/7346-batch-proving-circuits-and-l1.md
Comment (spalladino) [reply in-progress/7346-batch-proving-circuits-and-l1.md:215]: @LHerskind sounds good as a first approach, given that L1 gas is free!

File: in-progress/7346-batch-proving-circuits-and-l1.md
Comment (MirandaWood) [inline in-progress/7346-batch-proving-circuits-and-l1.md:165]:
```suggestion
- `left.block_number + 1 == right.block_number`
```
As discussed :)

**Comment by just-mitch** [comment]
@spalladino @MirandaWood @LeilaWang Is this one being implemented? Shall we merge the PR?

**Comment by spalladino** [comment]
@just-mitch Miranda has already implemented most of it. I was waiting on approval from @LeilaWang and @LHerskind, but I believe we can merge anyway.

**Comment by LeilaWang** [comment]
I'll look at the PR now!

**Review by LeilaWang — APPROVED** [review]

**Comment by just-mitch** [comment]
> Oh do you mean this doc pr? 😅

yep!

Sounds good! Thanks all!

**Pull request merged**
